### PR TITLE
Sized types for strong induction

### PR DESCRIPTION
**Pull request by ziman**
(This is not an implementation of [Agda's sized types](https://agda.readthedocs.io/en/v2.5.2/language/sized-types.html).)

This patch adds the proof of accessibility for values of any type that has a size mapping to `Nat`. For example, there is an implementation for lists, where `size = length`.

For any such type, we get strong induction on size, which allows recursive calls not only on direct subterms but on any value with smaller `size`.

### CODE REVIEW

**Comment by david-christiansen** [comment]
This is very nice!

Is `sizeAcc` the right name for it? I worry that `acc` in general will confuse people into thinking "accumulator". What if we picked "access" instead? I suppose it's too late to rename old Prelude things, but new ones?

What do you think?

File: libs/base/Data/List/Views.idr [hunk 1]
```diff
@@ -105,17 +101,17 @@ data Filtered : (a -> a -> Bool) -> List a -> Type where
...
+filteredROK : (p : a -> a -> Bool) -> (x : a) -> (xs : List a) -> Smaller (filter (\y => not (p y x)) xs) (x :: xs)
 filteredROK p x xs = LTESucc (filterSmaller xs)
 
 ||| Covering function for the `Filtered` view
 ||| Constructs the view in O(n lg n)
 export
 filtered : (p : a -> a -> Bool) -> (xs : List a) -> Filtered p xs
-filtered p inp with (smallerAcc inp)
+filtered p inp with (sizeAcc inp)
   filtered p [] | with_pat = FNil
```
Comment (ahmadsalim) [inline libs/base/Data/List/Views.idr]: I know it was there before, but I would prefer having `_` than `with_path` here.

**Comment by ziman** [comment]
Good ideas, thank you!

**Comment by ziman** [comment]
I went for `sizeAccessible` because the corresponding predicate is called `SizeAccessible` because the general predicate is called `Accessible`. The name `access` sounds better but I thought it might be safer to derive a name from the context.

Thoughts/suggestions?

**Comment by david-christiansen** [comment]
@ziman Sounds like a great choice.

**Comment by jfdm** [comment]
Looks good to me. I think a small CHANGELOG entry to highlight the new feature would be nice.

**Comment by ziman** [comment]
Thanks! CHANGELOG updated.

I also had to merge `master` because of a conflict in the CHANGELOG.

**Comment by ziman** [comment]
The failing tests have been fixed in `master` so hopefully this branch will be now marked with a green tick, too.

**Comment by melted** [comment]
Lets merge it before any CI breaks again :-)

**Pull request merged**